pkg/controller/controller_test.go

```diff
@@ -0,0 +1,238 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"fmt"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/latest"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/testclient"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/controller/framework"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+)
+
+// testControllerPlugin tests a Controller's ability to respond to watches on more than one object
+type testControllerPlugin struct {
+	host                     ControllerHost
+	pvSource                 *framework.FakeControllerSource
+	pvcSource                *framework.FakeControllerSource
+	expectedClaimAddedCount  int
+	expectedVolumeAddedCount int
+	actualClaimAddedCount    int
+	actualClaimDeletedCount  int
+	actualVolumeAddedCount   int
+	actualVolumeDeletedCount int
+	lock                     sync.Mutex
+	uids                     map[string]interface{}
+}
+
+func NewTestController(pvCount, pvcCount int) *testControllerPlugin {
+	return &testControllerPlugin{
+		expectedVolumeAddedCount: pvCount,
+		expectedClaimAddedCount:  pvCount,
+		actualClaimAddedCount:    0,
+		actualClaimDeletedCount:  0,
+		actualVolumeAddedCount:   0,
+		actualVolumeDeletedCount: 0,
+		uids: make(map[string]interface{}),
+	}
+}
+
+func (plugin *testControllerPlugin) Init(host ControllerHost) {
+	plugin.host = host
+}
+
+func (plugin *testControllerPlugin) Name() string {
+	return "test-controller-plugin"
+}
+
+func (plugin *testControllerPlugin) DefaultSyncPeriod() time.Duration {
+	return 1 * time.Second
+}
+
+func (plugin *testControllerPlugin) CreateController() Controller {
+	return plugin
+}
+
+func (plugin *testControllerPlugin) Configs() []*framework.Config {
+	// source simulates an apiserver object endpoint.
+	plugin.pvSource = framework.NewFakeControllerSource()
+	plugin.pvcSource = framework.NewFakeControllerSource()
+
+	// This will hold the downstream state, as we know it.
+	pvDownstream := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	pvcDownstream := cache.NewStore(cache.MetaNamespaceKeyFunc)
+
+	// This will hold incoming changes. Note how we pass downstream in as a
+	// KeyLister, that way resync operations will result in the correct set
+	// of update/delete deltas.
+	pvDeltaFifo := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, pvDownstream)
+	pvcDeltaFifo := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, pvcDownstream)
+
+	pvConfig := &framework.Config{
+		Queue:            pvDeltaFifo,
+		ListerWatcher:    plugin.pvSource,
+		ObjectType:       &api.PersistentVolume{},
+		FullResyncPeriod: 5 * time.Second,
+		RetryOnError:     true,
+		Process:          plugin.processPersistentVolume,
```

Comment:
I see us doing a lot of function pointers lately, and it seems mostly to be a way to dodge adding a typedef.  I am not the deepest in Go, but it seems that it would be more Goan (Goish?  Goy?) to do

```
package framework
type Config interface {
    GetQueue() whatever
    GetListerWatcher() whatever
    ...
    Process()
}

type StaticConfig struct {
    Queue 
    Listerwatcher
   ...
}

func (sc *StaticConfig) GetQueue() { return sc.Queue }
```

and

```
type myConfig struct {
    framework.StaticConfig
}

func (mc *myConfig) Process() { ...}
```


Upon typing that out, I realize how tedious it is, but I want to raise the point anyway.  There's an inflection point somewhere, whereupon adding function pointers is less readable that this mess..